### PR TITLE
*: support clustered index for fast analyze

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Problem Summary: 

```sql
drop table if exists t;
set @@tidb_enable_clustered_index=1;
set @@session.tidb_enable_fast_analyze=1;
create table t (a int, b int, c int, primary key(a, b));
insert into t values (1, 1, 1), (1, 2, 1), (1, 3, 1), (1, 4, 1), (1, 5, 1);
analyze table t;
```
```console
(1364, "Field 'a' doesn't have a default value")
```

Before this PR, `FAST ANALYZE` is not applicable in clustered index tables, because the key-value samples collected by `AnalyzeFastExec` are not processed in the case of the common handle.

What's Changed:

- Refactor
  - use `core.HandleCols` instead of `*model.ColumnInfo` to distinguish the primary key columns and to build collectors.
  - `DecodeRow` is renamed to `DecodeRowToDatumMap`, responsible for decoding a __row value__ to a datum map(column id -> datum).
  - export a new function `DecodeHandleToDatumMap`, responsible for decoding a __handle__ to a datum map.
- New
  - decode both the __key part__(handle columns) and the value part of a sample.
  - use `SampleStep` to scan from the kv storage, instead of choosing a random number.
- Ohter minor changes

### Related changes

N/A

### Check List 

Tests 

- Unit test
- Integration test

Side effects

N/A

### Release note 

- No release note

### CODE REVIEW

File: tablecodec/tablecodec.go [hunk 1]
```diff
@@ -367,11 +368,8 @@ func DecodeRowWithMapNew(b []byte, cols map[int64]*types.FieldType, loc *time.Lo
 		}
 		idx++
 	}
-	// for decodeToMap:
-	// - no need handle
-	// - no need get default value
```
Comment (coocood) [inline tablecodec/tablecodec.go:372]: Why remove the comments

Comment (tangenta) [reply tablecodec/tablecodec.go:372]: The parameters below are changed.

I added a comment at `DecodeRowToDatum`:
```go
// Default value columns, generated columns and handle columns are unprocessed.
func DecodeRowToDatum(b []byte, cols map[int64]*types.FieldType, loc *time.Location)
```

File: executor/analyze.go [hunk 2]
```diff
@@ -965,44 +980,28 @@ func (e *AnalyzeFastExec) handleSampTasks(bo *tikv.Backoffer, workID int, err *e
...
 		if task.SampSize == 0 {
 			continue
 		}
 
-		var tableID int64
-		var minRowID, maxRowID kv.Handle
-		startKey, endKey := task.Location.StartKey, task.Location.EndKey
-		tableID, minRowID, *err = tablecodec.DecodeRecordKey(startKey)
+		step := uint32((task.EndOffset - task.BeginOffset) / task.SampSize)
+		snapshot.SetOption(kv.SampleStep, step)
```
Comment (coocood) [inline executor/analyze.go:991]: We also need to set following options to minimize the cost on KV.
```
NotFillCache = true
IsolationLevel = kv.RC
Priority = kv.PriorityLow
```

Comment (coocood) [reply executor/analyze.go:991]: Well, `tikvSnapshot` doesn't have the `IsolationLevel` field but we can add one and use it to build scan request.

Comment (tangenta) [reply executor/analyze.go:991]: `pb.ScanRequest` does not have a field like `IsolationLevel`. How to wrap this message into the request?

**Comment by tangenta** [comment]
```
FAIL: session_test.go:3137: testSessionSuite3.TestPessimisticLockOnPartition

/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:207:
    tk.c.Assert(err, check.IsNil, check.Commentf("sql:%s, %v, error stack %v", sql, args, errors.ErrorStack(err)))
... value *errors.withMessage = previous statement: update forupdate_on_partition set first_name = 'xxx' where age=25: [domain:8028]Information schema is changed during the execution of the statement(for example, table definition may be updated by other DDL ran in parallel). If you see this error often, try increasing `tidb_max_delta_schema_count`. [try again later]
... sql:commit, [], error stack [domain:8028]Information schema is changed during the execution of the statement(for example, table definition may be updated by other DDL ran in parallel). If you see this error often, try increasing `tidb_max_delta_schema_count`. [try again later]
```

File: executor/analyze.go [hunk 3]
```diff
@@ -1160,30 +1160,8 @@ func (e *AnalyzeFastExec) buildStats() (hists []*statistics.Histogram, cms []*st
...
-	}
-	sort.Slice(randPos, func(i, j int) bool { return randPos[i] < randPos[j] })
-
-	for _, task := range e.sampTasks {
-		begin := sort.Search(len(randPos), func(i int) bool { return randPos[i] >= task.BeginOffset })
-		end := sort.Search(len(randPos), func(i int) bool { return randPos[i] >= task.EndOffset })
-		task.SampSize = uint64(end - begin)
-	}
+	totalSampSize := e.opts[ast.AnalyzeOptNumSamples]
+	e.estSampStep = uint32(e.rowCount / totalSampSize)
```
Comment (lzmhhh123) [inline executor/analyze.go:1164]: The `e.rowCount` is sum `mvcc.num_rows`. It contains many different mvcc version for one row. Do you make sure the `estSampStep` could handle this situation?

Comment (lzmhhh123) [reply executor/analyze.go:1164]: By the way, the `estSampStep` filters certain rows for a not changing table. Could we add some rand for it?

File: executor/analyze.go [hunk 4]
```diff
@@ -962,47 +982,30 @@ func (e *AnalyzeFastExec) handleSampTasks(bo *tikv.Backoffer, workID int, err *e
 	if *err != nil {
 		return
 	}
+	snapshot.SetOption(kv.NotFillCache, true)
+	snapshot.SetOption(kv.IsolationLevel, kv.RC)
+	snapshot.SetOption(kv.Priority, kv.PriorityLow)
```
Comment (lzmhhh123) [inline executor/analyze.go:987]: Why we need these?

File: executor/analyze.go [hunk 5]
```diff
@@ -537,11 +549,21 @@ func (e *AnalyzeColumnsExec) buildStats(ranges []*ranger.Range) (hists []*statis
 	return hists, cms, nil
 }
 
+func hasPkHist(handleCols core.HandleCols) bool {
+	return handleCols != nil && handleCols.IsInt()
```
[the remark this answers is not in the document]
Comment (tangenta) [reply executor/analyze.go:553]: Yes. Only when the pk is handle, an `IntHandleCols` can be created. See the function `BuildHandleColsForAnalyze`.

File: executor/analyze.go (hunk 4 above)
Comment (tangenta) [reply executor/analyze.go:987]: These are optimizations.

`NotFillCache` & `PriorityLow`: the sampling task should not affect the normal kv operations.
`Read Committed Isolation`: it speeds up the kv scan operation.

Comment (lzmhhh123) [reply executor/analyze.go:987]: OK.

File: executor/analyze.go (hunk 3 above)
Comment (tangenta) [reply executor/analyze.go:1164]: > The `e.rowCount` is sum `mvcc.num_rows`. It contains many different mvcc version for one row. Do you make sure the `estSampStep` could handle this situation?

@lzmhhh123 @coocood Do you have any suggestions to deal with this problem?

File: executor/analyze.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (lzmhhh123) [reply executor/analyze.go:553]: OK.

File: executor/analyze.go (hunk 3 above)
Comment (lzmhhh123) [reply executor/analyze.go:1164]: Maybe we can make `estSampStep` scaling by `region.mvcc.num_rows / region.num_rows` on tikv?

Comment (coocood) [reply executor/analyze.go:1164]: The `mvcc.num_rows` only contains one version for each key.

Comment (coocood) [reply executor/analyze.go:1164]: `mvcc.num_versions` constains all the versions in the region.

File: tablecodec/tablecodec.go [hunk 6]
```diff
@@ -445,39 +445,40 @@ func DecodeHandleToDatumMap(handle kv.Handle, handleColIDs []int64,
...
+				return row, err
+			}
+			row[id] = d
+			break
+		}
+	}
+	return row, nil
+}
+
+func decodeHandleToDatum(handle kv.Handle, ft *types.FieldType, idx int) (types.Datum, error) {
```
Comment (coocood) [inline tablecodec/tablecodec.go:468]: The `ft` is not used.

File: executor/analyze.go [hunk 7]
```diff
@@ -1111,20 +1056,21 @@ func (e *AnalyzeFastExec) runTasks() ([]*statistics.Histogram, []*statistics.CMS
...
 		if len(collector.Samples) > 0 {
 			collector.TotalSize *= rowCount / int64(len(collector.Samples))
 		}
-		if i < hasPKInfo {
-			hists[i], cms[i], err = e.buildColumnStats(e.pkInfo.ID, e.collectors[i], &e.pkInfo.FieldType, rowCount)
-		} else if i < hasPKInfo+len(e.colsInfo) {
-			hists[i], cms[i], err = e.buildColumnStats(e.colsInfo[i-hasPKInfo].ID, e.collectors[i], &e.colsInfo[i-hasPKInfo].FieldType, rowCount)
+		if i < pkColCount {
+			pkCol := e.handleCols.GetCol(i)
+			hists[i], cms[i], err = e.buildColumnStats(pkCol.ID, e.collectors[i], pkCol.RetType, rowCount)
```
Comment (lzmhhh123) [inline executor/analyze.go:1069]: `rowCount` here is calculated by `e.rowCount` which is sum `(sampleStep+1) * sampleCount + scanCount`. Why not directly use the result in `calculateEstimateSampleStep`.

File: executor/analyze_test.go [hunk 8]
```diff
@@ -516,7 +549,7 @@ func (s *testFastAnalyze) TestFastAnalyzeRetryRowCount(c *C) {
 	tk.MustQuery("select * from retry_row_count")
 	tk.MustExec("analyze table retry_row_count")
 	// 4 regions will be sampled, and it will retry the last failed region.
-	c.Assert(cli.mu.count, Equals, int64(5))
+	c.Assert(cli.mu.count, Equals, int64(0))
```
Comment (coocood) [inline executor/analyze_test.go:552]: We can remove this line.

File: executor/analyze.go (hunk 7 above)
Comment (tangenta) [reply executor/analyze.go:1085]: Updated. Now the row count in `stats_meta` won't be updated during fast analyze, unless it is the first time.

File: executor/analyze_test.go [hunk 9]
```diff
@@ -234,14 +264,15 @@ func (s *testFastAnalyze) TestAnalyzeFastSample(c *C) {
...
-		PKInfo:          pkCol,
+		HandleCols:      handleCols,
 		ColsInfo:        colsInfo,
 		IdxsInfo:        indicesInfo,
 		Concurrency:     1,
 		PhysicalTableID: tbl.(table.PhysicalTable).GetPhysicalID(),
 		TblInfo:         tblInfo,
 		Opts:            opts,
 	}
+	tk.MustExec("begin")
```
Comment (lzmhhh123) [inline executor/analyze_test.go:275]: Why need the transaction?

Comment (tangenta) [reply executor/analyze_test.go:275]: This is used in `calculateEstimateSampleStep`:
```go
txn, err := e.ctx.Txn(true)
if err != nil {
	return 0, errors.Trace(err)
}
txn.SetOption(kv.Priority, kv.PriorityLow)
txn.SetOption(kv.IsolationLevel, kv.RC)
txn.SetOption(kv.NotFillCache, true)
sql := fmt.Sprintf("select count(*) from %s.%s;", dbInfo.Name.L, e.tblInfo.Name.L)
recordSets, err := e.ctx.(sqlexec.SQLExecutor).ExecuteInternal(context.TODO(), sql)
```

Comment (tangenta) [reply executor/analyze_test.go:275]: The test here invokes the method `buildStats()` in the fast-analyze executor directly. If we don't `begin`, it reports 'invalid transaction'.

Comment (tangenta) [reply executor/analyze_test.go]: Updated. Now it will run 'begin/rollback' internally if the transaction is invalid.

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
[2020-08-07T11:17:51.802Z] FAIL: merge_join_test.go:244: testSerialSuite1.TestMergeJoinInDisk
[2020-08-07T11:17:51.802Z] 
[2020-08-07T11:17:51.802Z] merge_join_test.go:277:
[2020-08-07T11:17:51.802Z]     c.Assert(tk.Se.GetSessionVars().StmtCtx.DiskTracker.MaxConsumed(), Greater, int64(0))
[2020-08-07T11:17:51.802Z] ... compare_one int64 = 0
[2020-08-07T11:17:51.802Z] ... compare_two int64 = 0

**Pull request merged**